src/object/obj_tx.c

```diff
@@ -864,19 +987,14 @@ dc_tx_restart(tse_task_t *task)
 		rc = -DER_NO_PERM;
 	} else {
 		dc_tx_cleanup(tx);
-		tx->tx_status = TX_OPEN;
 
-		/*
-		 * Increasing the epoch doesn't change our knowledge of its
-		 * uncertainty on the client side.
-		 *
-		 * FIXME: The tx_epoch should be reset as zero, but before
-		 *	  server side HLC based solution ready, we temporarily
-		 *	  set it as client side HLC.
-		 *
-		 *	tx->tx_epoch = 0;
-		 */
-		tx->tx_epoch.oe_value = crt_hlc_get();
+		tx->tx_status = TX_OPEN;
+		if (dtx_epoch_chosen(&tx->tx_epoch))
+			tx->tx_epoch.oe_flags |= DTX_EPOCH_HINT;
```

Comment:
dc_tx_restart() may be triggered by the TX sponsor if some DAOS API (such as fetch) return "-DER_TX_RESTART", but at that point, the fetch RPC reply may already contained the valid epoch and set to the tx_epoch via dc_tx_set_epoch(). So here, if we mark the tx_epoch.oe_flags as "DTX_EPOCH_HINT", does it means we need another fetch to refetch another epoch from the server after the restart?

Reply:
Yes, the restarted TX begins without a chosen epoch. If the user calls FETCH first, this FETCH chooses an epoch; if the user calls UPDATE first, the CPD chooses an epoch, etc. The HINT flag ensures that any new epoch chosen will be higher than the previous chosen epoch.